### ISSUE TITLE
[RFC] User kind location

### ISSUE THREAD

**Issue by ProticM** [backstage/backstage#7449]
**Status:** Open for comments

## Need

Currently, all provided examples regarding the `Group` kinds (and users related to it) location is such that each group is in the single file together with the users that are members of that particular group. This is perfectly fine in some scenarios, but the problem comes with the situation when a single user belongs to multiple groups/teams.

I'm aware that in terms of recognizing these users in the portal we do not have an issue, but it becomes unlogical to place the `user-a` in the `team-a` YAML file if that user is also a member of `team-b`.

I'm looking for the best practice approach for the situation above. What would be a proposed solution regarding this topic?

## Proposal

IMHO the solution for the issue above would be to extract the users from the group YAML file, create each user in a separate file under, let's say, the `users` folder, and create a `Location` kind pointing to all of them as targets. Then, we simply register all the users via that location. I'm aware that we lose the grouping, and when editing the group we do not see the users in the file, but in my case extracting the users will bring more benefits than harm.

It is cleaner to create users per file, especially that we are a big organization with over 3000 users to be registered.

## Risks

Are 3000 users too much to be handled by a single `Location` kind?

**Comment by freben** [comment]
I'll start by saying that the examples are just that - examples just to have some data in the example deployment. They are very small scale and won't work nicely in the same arrangement for most real life scenarios.

Most adopters ingest their users and groups from an external source, for example LDAP, rather than using YAML files. See the LDAP org data [documentation](https://backstage.io/docs/integrations/ldap/org) and the other nearby "Org Data" entries in the sidebar.

On the topic of addressing a large number of files, most of our readers actually support globs. You could register `https://github.com/myorg/myrepo/orgdata/**/*.yaml` for example as a single location, and it would automatically address every single file it finds that matches that pattern. This has benefits and drawbacks, but might be interesting to know about while you set yourselves up for automatic ingestion from an authority.

**Comment by freben** [comment]
> Yes, we are also discussing, in parallel to this conversation, how to incorporate an external source by creating a custom data reader processor for the users.

Are you not in the position that you have a stable and settled org hierarchy yet? That's actually pretty common.

> Regarding LDAP I must check for that because I'm a new person in the organization. Also, regarding globs, it is an interesting thing to know, but a `Location` kind makes more sense to me. Can it handle a large number of targets?

Yes, I do not foresee any problems with referring to a larger number of individual targets. Just that the Location file gets pretty large and unwieldy to edit. But sounds like you may have tooling in place that helps with that.

> At the moment, we are missing some information in the current organizational structure that we would like to display in the backstage portal thus I created this RFC here and due to that, we are considering creating the YAML files.

It could be that you end up importing the things you _do_ have in an automated fashion, and then just "fill in the blanks" with YAML files so that they aren't so many. It's fine to have several sources of data, as long as they don't try to overwrite each other.

> If we register from an external source, how does backstage handle new additions? For example, if we onboard a new person, how is that single entity handled by the custom data reader? I assume that in that case, we update our existing structure and the catalog will look in that external location from time to time and update the entities?

Exactly. That processor would poll your external source at some regular cadence, and it would emit all of the resulting entities over and over again. Anything new that appears in the set of emitted things, will just be persisted and appears as a new entity.

> Does this mean that we do not use backstage to handle the creation of the new users? If so, this is a bit of a limitation in terms of template usages, because we plan to have a template in backstage to create a new entity kind (User, API, System...) in our repos.

Remember that Backstage's primary use case is to be a mirror into already existing things, rather than replacing existing things.

Let's take our own company as an example. Here, we have decided to have a microservice as the ultimate authority for users and groups, and that service can also handle creation and deletion. Some team owns and manages that service, completely separate from Backstage. It could just as well have been LDAP directly, or some other solution, doesn't matter.

Then, what has been added to Backstage is,

- A (closed source, internal because it's tied to that specific closed source service) frontend plugin that contains some nice views for how to create, edit and manage users, and
- An ingestion of users and groups into the catalog

So we keep the original source intact, we don't have any hard ties to Backstage (you could just as well manage users using any other means, even `curl`), but we make it so that there's an excellent end user experience for doing these management tasks. You could to the same.

> In other words, it would be a certain amount of limitation in scaffolding a new user from a template into our existing external location for the users (without fully customized actions)

Yeah you could manage the users in any way you see fit, and then the idea is that the catalog just mirrors that external source, whatever it is.

**Comment by freben** [comment]
Right, start out like this and then see where it takes you. It just feels like it risks becoming a case of "reinventing active directory or similar product, but more bespoke and fewer features and guarantees" :) I see where you are going with it, and sometimes it can be enough. Just that, you know, there's compliance and legal and practical matters of all kinds that may come into play when it comes to org management, and just be ready to switch to a dedicated prefab solution to ingest from when you're outgrowing what you have.

**Comment by freben** [comment]
Yeah we all live in a reality and things aren't always yet in their dream state! No worries.

Feel free to close the issue or continue discussion as you see fit.

**Comment by freben** [comment]
No, there is no merging. Having a duplicate entity will just lead to a conflict such that the catalog refuses to even consider accepting the second one. This is a security feature to prevent takeovers.

Luckly, groups can have members, and users can be memberOf. So you can express things from either end with the same outcome.

**Comment by ProticM** [comment]
I understand, but in that case, I do not see a way how to link the user and a group if both the entities are already registered in the catalog (from the external source) without having `members` or `memberOf` filled in 🤔

**Comment by ProticM** [comment]
> Ah right. I thought you could have an autoritative source for for example 90% of the groups, but then wanted to add the remaining 10% as yaml files to for example fill in an org hierarchy. Or to add users that aren't yet onboarded but still are members of groups.

This part is clear, but it is not my case unfortunately.

> But OK, if you have all groups and users to import but not all of their relations are importable or knowable from that system

Kinda 😄 For groups and users (mostly for users), I have some options, but the source data for other kinds is an excel sheet

> ...add a custom processor to your catalog, which has the ability to fill in those edges on the fly...

Yes, this is the final solution I'm going for after gathering information and discussing various approaches. The user data that we have in the active directory and in our HR system will be ingested, the missing gaps will be added on the fly during the process of parsing that raw data in my custom processor.

> ...like a hand written yaml/csv/whatnot file on a repo somewhere or similar...

This is why we have our templates to help us write those YAML files as described in my previous comments.

Ideally, I would love to create a separate microservice and manage all my kinds in it with a help of a frontend application, and then expose API per kind to be ingested by the backstage...at the moment, it is not an option
